src/Lachain.Core/Consensus/EraBroadcaster.cs

```diff
@@ -190,7 +279,9 @@ public void Dispatch(ConsensusMessage message, int from)
                     throw new InvalidOperationException($"Unknown message type {message}");
             }
 
-            HandleExternalMessage(protocolId, new MessageEnvelope(message, from));
+            var messageEnvelope = new MessageEnvelope(message, from);
+            _messageEnvelopeRepositoryManager.AddMessage(messageEnvelope);
```

Comment:
Consider moving this line inside HandleExternalMessage method and persist if protocolId is validated. Also try to find a way to check if the message was from restore or not, if from restore then don't need to persist again.